unit-tests/src/org/commcare/android/database/user/models/SessionStateDescriptorTests.java

```diff
@@ -0,0 +1,47 @@
+package org.commcare.android.database.user.models;
+
+import org.commcare.models.AndroidSessionWrapper;
+import org.commcare.modern.session.SessionWrapper;
+import org.commcare.session.CommCareSession;
+import org.commcare.test.utilities.MockApp;
+import org.commcare.util.CommCarePlatform;
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.javarosa.core.model.instance.TreeElement;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class SessionStateDescriptorTests {
+
+    @Test
+    public void testSessionStateDescriptorSerialization() throws Exception {
+        MockApp app = new MockApp("/commcare-apps/case_search_and_claim/");
+        SessionWrapper session = app.getSession();
+        session.setCommand("patient-search");
+        session.setQueryDatum(ExternalDataInstance.buildFromRemote("foo", new TreeElement()));
+        session.setDatum("case_id", "case_two");
+        serializeSessionOutToDescriptor(session);
+
+        session.clearAllState();
+        session.setCommand("m0");
+        session.setCommand("m0-f0");
+        session.setDatum("case_id", "case_two");
+        session.setComputedDatum(session.getEvaluationContext());
+
+        serializeSessionOutToDescriptor(session);
+    }
+
+    private static void serializeSessionOutToDescriptor(CommCareSession session) {
```

Comment:
It tests 
 - A: building session descriptors from session wrappers
 - B: building session wrappers from session descriptors.
It makes sure that doing A(B(A(session))) gives you back the same session descriptor that you started with.